Tutorial for odc-stac

caitlinadams opened the pull request:
### Reason for this pull request

The documentation subcommittee is working to add tutorials to the Open Data Cube docs website. This is the first of these tutorials, which shows how to load data using odc-stac.

The design pattern is that a full walkthrough is available in the docs page, and this is supported by notebooks held in a [companion repository](http://github.com/opendatacube/tutorial-odc-stac). This allows users to easily launch a binder instance and complete the tutorial. The companion repository is independently documented so that people can run it if accessing the tutorial via GitHub rather than the docs website.

### Proposed changes

- Add a tutorials folder to docs
- Add an odc-stac tutorial document, which has a [companion repository](http://github.com/opendatacube/tutorial-odc-stac)

<!-- readthedocs-preview opendatacube start -->
----
📚 Documentation preview 📚: https://opendatacube--2055.org.readthedocs.build/en/2055/

<!-- readthedocs-preview opendatacube end -->

caitlinadams committed:
Create initial tutorial pages

caitlinadams committed:
Merge branch 'develop' into doc-tutorial-odcstac

caitlinadams committed:
Add introduction and perparation sections

caitlinadams committed:
merge develop

caitlinadams committed:
Add links for git and conda to odc-stac tutorial

caitlinadams committed:
Merge branch 'develop' into doc-tutorial-odcstac

caitlinadams committed:
Add instructions for virtual environment

caitlinadams committed:
Add instructions to launch binder and headings for tutorial

caitlinadams committed:
Merge branch 'develop' into doc-tutorial-odcstac

caitlinadams committed:
update text about launching environment

caitlinadams committed:
Merge branch 'develop' into doc-tutorial-odcstac

caitlinadams committed:
Continue updating tutorial

caitlinadams committed:
Start updating tutorial to use IO LULC

caitlinadams committed:
Add content for STAC concepts

caitlinadams committed:
Add text on searching and loading data

caitlinadams committed:
Major update to complete final sections and tidy

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

caitlinadams committed:
Fix badly formatted links

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

caitlinadams committed:
Add non-standard words to wordlist

caitlinadams committed:
Merge branch 'develop' into doc-tutorial-odcstac

pjonsson commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,290 @@
+Many Earth observation data providers generate STAC metadata, which can be used to find and load data you're interested in.
+STAC metadata has four important components:
+
+* **Catalog**: A structure for organising multiple datasets managed by a given provider. For example, `Planetary Computer's Catalog <pc-stac_>`_
```
This seems to add new warnings to the doc build (https://app.readthedocs.org/api/v2/build/28897729.txt):
```
/home/docs/checkouts/readthedocs.org/user_builds/opendatacube/checkouts/2055/docs/tutorials/odc-stac.rst:133: ERROR: Indirect hyperlink target "planetary computer's catalog"  refers to target "pc-stac", which does not exist. [docutils]
/home/docs/checkouts/readthedocs.org/user_builds/opendatacube/checkouts/2055/docs/tutorials/odc-stac.rst:134: ERROR: Indirect hyperlink target "land use land cover collection"  refers to target "pc-lulc", which does not exist. [docutils]
/home/docs/checkouts/readthedocs.org/user_builds/opendatacube/checkouts/2055/docs/tutorials/odc-stac.rst:135: ERROR: Indirect hyperlink target "land use land cover data for supercell 15m in 2013"  refers to target "pc-item", which does not exist. [docutils]
/home/docs/checkouts/readthedocs.org/user_builds/opendatacube/checkouts/2055/docs/tutorials/odc-stac.rst:133: ERROR: Unknown target name: "pc-stac". [docutils]
/home/docs/checkouts/readthedocs.org/user_builds/opendatacube/checkouts/2055/docs/tutorials/odc-stac.rst:134: ERROR: Unknown target name: "pc-lulc". [docutils]
/home/docs/checkouts/readthedocs.org/user_builds/opendatacube/checkouts/2055/docs/tutorials/odc-stac.rst:135: ERROR: Unknown target name: "pc-item". [docutils]
```

SpacemanPaul replied:
Yes. All these `pc-*` link labels appear to be undefined.

caitlinadams replied:
Apologies, I mistakenly deleted these when cleaning up other URLs that docs was complaining about. I will add them back in.

pjonsson commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,290 @@
+
+.. code-block:: python
+
+   catalog_query = "https://planetarycomputer.microsoft.com/api/stac/v1/"
```
The naming convention for variables in this tutorial seems to roughly be `<name>_<type>`, and this does not look like anything that I could associate with a query. I would naively call this thing an endpoint, and I would call the two following lines selectors.

robbibt replied:
I think just "catalog" would work well here, instead of "catalog_query"

caitlinadams replied:
Yes, happy to make a change. I like to use a suffix of some kind to help the user distinguish between named key-word arguments in functions and the values that are being supplied, but I agree that query is not such a good suffix.  In pystac_client's documentation, this arg is called "id" so I think naming this variable "catalog_id" or "catalog_url" makes sense

pjonsson replied:
If the target group for this tutorial isn't programmers, I think you have a strong pedagogical point in using variable names that differ from the argument names to prevent the misconception that there is some kind of teleportation between defining a variable with a specific name and the function magically knowing what to do.

On a conceptual level, I feel that with the particular names chosen, the situation resembles this example:
```python
first_number_multiplication = 3
second_number_multiplication = 5
```
where neither of the defined things are multiplications. Just like chucking in those two variables as arguments to `*` makes the compound `first_number_multiplication * second_number_multiplication` become a multiplication, the defined `_query` variables don't become a query until they're chucked together to some function/method somewhere inside `search()`.

After writing the previous paragraph, I reflected that flipping the words around and calling the variable `multiplication_first_number` instead makes the situation feel different to me, so for some reason I'm not expecting `multiplication_first_number` to hold the result of a multiplication. I can't articulate why though, it could be conventions used in a lot of code bases, or some bias caused by me only knowing Germanic languages.

caitlinadams replied:
It's definitely interesting that changing the order of words in a variable makes such a difference! I've decided to change a couple of the variable names to be more descriptive and readable, i.e. `desired_collections` rather than `collection_query`. This still achieves my aim of showing the user that there's no magic link between key-word args and variable names, along with more descriptive variable names. Thanks for the feedback on this one!

pjonsson replied:
They say there are only two hard problems in computer science: cache invalidation, naming things, and off-by-1 errors!

I think desired is a great prefix, it conveys the intent without distorting the underlying thing!

pjonsson commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,290 @@
+
+   start_date = "2017-01-01"
+   end_date = "2023-01-01"
+   date_query = (start_date, end_date)
```
The heading is "Date range" and the way I read the text implies a date range, so those two seem aligned. Continuing from that buildup into the code and then not naming the variable `date_range` is a bit surprising to me.

caitlinadams replied:
I like using the suffix `query` to help users distinguish between named kwargs in the query/loading functions later in the notebook. I can rename it to `date_range_query` to better align with the title and text.

caitlinadams replied:
Have changed this to `date_range` now

pjonsson commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,290 @@
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
+You may see warnings appear in red text.
+This is expected behaviour for this dataset and does not mean anything is wrong.
```
What general conclusions do you want a reader to have drawn after reading this sentence?

I'm a bit concerned that most readers will draw the conclusion that warnings when loading a dataset is expected, so warnings are just noise that could/should be ignored. Short-term this doesn't matter, but in the long run, I think users learning that lesson will get bitten by ignoring warnings that do matter. (I assume the warnings aren't bogus, but if they are, we should stop emitting them.)

robbibt replied:
Many beginner users tend to get very concerned by warnings - in training sessions I've seen them lead users to abandon tutorials early as they assume they are full errors that indicate they've done something wrong. So I think this message is primarily aimed at that class of user, rather than more advanced users who would gain more from actually reading the warnings.

ODC does tend to produce many "useless" warnings on data load (e.g. warnings that appear to have no impact on the result, but clutter up the output). I would love it if we could clean those up or work out what causes them.

<img width="749" height="106" alt="image" src="https://github.com/user-attachments/assets/2168d511-4027-4985-8691-7cddf9e369f3" />

caitlinadams replied:
Robbi has it right -- unfortunately, the load command produces a very unintelligible warning for this dataset. It's being triggered by `affine`, which is not a library we can control, nor do we control the source dataset. The only alternative I can see would be to suppress the warning, but I'm not sure that's a good thing to teach people either? I can include an image of the warning so that users aren't panicked by it -- that's really all I'm trying to avoid. I can change the language to say the user shouldn't worry about seeing this specific RunTimeWarning, but I don't want to overload them. 

It's mostly hard because warnings show up in a red box in notebooks, which looks much more like an Error than a Warning to novice users.

<img width="1017" height="610" alt="Screenshot 2025-07-21 at 4 16 56 pm" src="https://github.com/user-attachments/assets/aeb5ac28-fe68-49a1-b5da-aaf9573c3c2a" />

pjonsson replied:
I'm not sure if I qualify as a beginner user, but the pictures scare the crap out of me.

The Rasterio warning seems like something ODC could suppress if the dataset doesn't have any of the mentioned things since it's an expected warning, but why is a reprojection happening in the first place if there isn't a transformation matrix, can ODC just avoid reprojecting when all these things are missing?

I have seen a bunch of warnings concerning Affine in PyCharm, are they related to the RuntimeWarning? (The `_base.py` warning is in line 1330.)

<img width="285" height="138" alt="image" src="https://github.com/user-attachments/assets/a7fb2144-1c86-47f0-9402-14484c72218b" />

Edit: I know more recent Python versions have had improvements to the error messages omitted, and I imagine that is beneficial for novice users, but I notice both your screenshots are using Python 3.10. Is there some reason for the tutorial to remain on the oldest Python version possible? (Not suggesting that affects the warnings being discussed here, just wondering if users could be helped even more "for free" by switching to a more recent Python version.)

robbibt replied:
@kirill888 @SpacemanPaul Have either of you seen the warnings in Caitlin's example? This feels pretty excessive and something that significantly affects user experience - would be great to work out what's going wrong here.

@CaitlinAdams Are the warnings random, or easy to reproduce?

caitlinadams replied:
Looks like I'm using Python 3.10 in Binder because that's the default. I can update this to Python 3.13. However, I've tried running the notebook in Python 3.13 on my local machine and the warning is still produced. 

@robbibt -- the warning is somewhat reproducible. I usually get a similar number of warnings as items (the example loads 17 STAC items, and produces 14 warnings). If I load the items one at a time, I get warnings for some items, but not others. If I load items I didn't get warnings for as a group, I suddenly get warnings again. Shrug.

Having done a bit of looking, it seems I can use the warnings package to rule out warnings from affine specifically using 
```
import warnings
warnings.filterwarnings('ignore', module="affine")
```

I think I can explain to the user that we need this command for this specific dataset, and that it's safe to ignore the specific warning from the affine module. @pjonsson -- do you think this is a reasonable compromise between not suppressing warnings and saving the user from scary red text?

caitlinadams replied:
I've updated the wording in the document to specifically point out that we expect RunTimeWarnings from affine for this dataset. I'd still be interested in thoughts on adding code to selectively filter the affine warnings.

robbibt replied:
I think suppressing warnings in this demo notebook is definitely good idea... but it feels like something we want to resolve at the source in `odc-stac`, too - the current output really gives the impression something has gone badly wrong, which is misleading if that's not the case.

Kirill888 replied:
@caitlinadams @robbibt 

I had a look, error happens when trying to compute item footprint in output pixel coordinate frame. Reason for it: can't transform world coordinates from UTM60S to UTM15S, which is not surprising. STAC search end point returns items that do not overlap with the query, which is in UTM15S.

```python
>>> [it.properties["proj:code"] for it in items]
['EPSG:32760',
 'EPSG:32715',
 'EPSG:32701',
 'EPSG:32760',
 'EPSG:32715',
 'EPSG:32701',
 'EPSG:32760',
 'EPSG:32715',
 'EPSG:32760',
 'EPSG:32715',
 'EPSG:32760',
 'EPSG:32715',
 'EPSG:32701',
 'EPSG:32760',
 'EPSG:32715',
 'EPSG:32760',
 'EPSG:32715']
```

`odc-geo` not handling the case when coordinate transforms generate infinite values when computing block overlap.

FYI: I'm currently not paid by any org to work on ODC, or on anything, actually, so probably won't be engaged a lot in here.

pjonsson replied:
> Looks like I'm using Python 3.10 in Binder because that's the default. I can update this to Python 3.13. However, I've tried running the notebook in Python 3.13 on my local machine and the warning is still produced.

Sorry for being unclear, the warnings are orthogonal to the Python version comment. My train of thought was that beginners are more likely to mistype the tutorial and get unrelated error messages since they have no mental model of what Python code should look like. But since more recent versions of Python has improved error messages, that might save the user from at least some confusion.

> ```python
> import warnings
> warnings.filterwarnings('ignore', module="affine")
> ```
> 
> I think I can explain to the user that we need this command for this specific dataset, and that it's safe to ignore the specific warning from the affine module. @pjonsson -- do you think this is a reasonable compromise between not suppressing warnings and saving the user from scary red text?

(You get syntax highlighting if you add `python` right after the 3 opening backticks for the code block.)

If the picture from your earlier comment is what we're trying to avoid, you should be able to narrow the filter even more by adding the `messages` argument to `filterwarnings`. If that works it's a pretty surgical suppression, and I don't see much harm even if the line spreads like wildfire.

I do agree with Robbi that it would be preferable to fix the underlying problem though, both for the user experience in general, and so the tutorial can focus on the subject without going off on tangents about warnings.

Kirill888 replied:
@pjonsson error comes from `odc-geo` code responsible for overlap computation, triggered by items returned from stac end-point which returns items far away from the query region, even though it is supposed to filter those items out.

Kirill888 replied:
Basically, items that span `lon=0` are chopped into multi-polygons, but incorrectly, as a result they get returned from every query that has the same lat range.

<img width="642" height="352" alt="image" src="https://github.com/user-attachments/assets/99018195-2018-4ac4-b083-33334ce5aabe" />

Example of broken geometry:

https://radiantearth.github.io/stac-browser/#/external/planetarycomputer.microsoft.com/api/stac/v1/collections/io-lulc-annual-v02/items/60M-2023?.language=en

Kirill888 replied:
> ODC does tend to produce many "useless" warnings on data load (e.g. warnings that appear to have no impact on the result, but clutter up the output). I would love it if we could clean those up or work out what causes them.
> <img alt="image" width="749" height="106" src="https://private-user-images.githubusercontent.com/17680388/468476270-2168d511-4027-4985-8691-7cddf9e369f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTMxODAzNDYsIm5iZiI6MTc1MzE4MDA0NiwicGF0aCI6Ii8xNzY4MDM4OC80Njg0NzYyNzAtMjE2OGQ1MTEtNDAyNy00OTg1LTg2OTEtN2NkZGY5ZTM2OWYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzIyVDEwMjcyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4NGVjNmQxOGU3Y2FmMmE5NDNkZjRlNTI3YWZjNWYzNDJhZmU5ZWEyNDA1MWQwNDAzOTM4ZjljYWQyNjdjMWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1WDOsvicBPg6db4T2yMUTSxU4tmwBn4B-JDubfkQHQg">

https://github.com/opendatacube/odc-geo/issues/131

deep inside rasterio, still no clue what exact circumstance triggers it, we already have some filters installed to hide those, but not everthing is caught.

pjonsson replied:
> triggered by items returned from stac end-point which returns items far away from the query region, even though it is supposed to filter those items out.

@Kirill888 odc-geo not accounting for inconsistent data sounds reasonable. Do you know if the wacky data returned from the search is a server issue, faulty content of the underlying STAC files, or a bug in pystac?

Kirill888 replied:
Broken footprint stored on ms servers. Someone should raise an issue there, see links to a broken item in my comment above

caitlinadams replied:
Thanks for digging into this Kirill, I really appreciate it. I've logged an issue: https://github.com/microsoft/PlanetaryComputer/issues/454

pjonsson replied:
Is there some external deadline (scheduled workshop/hackathon/etc.) for when this tutorial needs to be merged, or can it sit for a few days? I don't have any experience with PlanetaryComputer, but some software is amazing and fixes stuff within hours after it is reported, so I'm still hopeful the tutorial can avoid having to go into the tangent about the warnings.

caitlinadams replied:
No, it's not critical -- I mostly just want it finished so I can get it off my to-do list! I'm not optimistic about Planetary Computer addressing it -- they were quite active in replying to people's issues in the repo up until early June and it seems to have gone quiet since then.

The easiest fix, I think, would be to switch to an AOI that's not near the equator or the antimeridian. I picked Galapagos to match the other tutorial we're developing, but it's not critical. So I might change the AOI in the next few days and see if I can do an area without warnings.

caitlinadams replied:
I've changed the AOI to one of the Canary Islands, which avoids the warnings. I've removed the text about the user needing to be aware of the warnings.

SpacemanPaul commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,290 @@
+
+Once you have loaded and checked your data, it is often useful to export it.
+This allows you to use the data in other software and analyses.
+The :code:`odc-geo` package adds the `.odc` extension for `xarray` datasets, allowing easy exporting of Cloud Optimised GeoTIFFs.
```
Not sure why this is mentioned when the code below doesn't seem to use the `.odc` extension?

caitlinadams replied:
It's a hangover from when I was previously using it, which I cleaned up in the notebook, but not the .rst. Will fix.

SpacemanPaul reviewed (changes requested):
Mostly looks good.  Some broken links to fix and some potential cleanup for clearer communication.

robbibt commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,290 @@
+   
+       ds_single_year = ds.data.isel(time=timestep)
+       
+       write_cog(
```
I think we should use the `.odc.write_cog()` extension instead of importing the function itself. It's a really nice way to interact with ODC data, and once users are familiar with it there's a lot of handy tools there they can use.

Would also fit with the sentence above: `adds the `.odc` extension for `xarray` datasets, allowing easy exporting of Cloud Optimised GeoTIFFs.`

caitlinadams replied:
I think the text is a hang-over of when I was using the .odc extension, which I started doing because I also liked it, but then worried that it might be one thing too many to explain to the user. 

What would you think of us writing another tutorial to highlight the useful functionality of odc-geo and the .odc extension?

caitlinadams replied:
I think the text is a hang-over of when I was using the .odc extension, which I started doing because I also liked it, but then worried that it might be one thing too many to explain to the user. 

What would you think of us writing another tutorial to highlight the useful functionality of odc-geo and the .odc extension?

robbibt commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,290 @@
+
+.. code-block:: python
+
+   ds.data.plot.imshow(col="time", col_wrap=3)
```
Why the `.data` here? Is this to force computation of the data prior to plotting it? (it's not usually necessary).

Should we include an example of using `.odc.explore()` to quickly plot an interactive map? e.g. `ds.red.isel(time=0).odc.explore()`

caitlinadams replied:
Ah, that's because `data` is unfortunately the name of our asset, such that `ds` is an xarray.DataSet and `ds.data` is an xarray.DataArray.

I could perhaps make this clearer by doing `ds["data"]`. What do you think?

caitlinadams replied:
Also, I tried using .odc.explore(), but the colours ended up really funky for displaying landcover, and it doesn't make a lot of sense without a legend (which is too much code to add for a simple tutorial). I think explore is nice for satellite imagery, same with converting to RGBA, so I would suggest we also cover that in an odc-geo specific tutorial. Very happy to hear your thoughts though!

robbibt reviewed (commented):
This looks amazing @caitlinadams! A few minor comments below, mainly about featuring the `.odc.*` extension a little more as I think it's one of the nicest parts about working with ODC. 🙂

caitlinadams committed:
Merge branch 'develop' into doc-tutorial-odcstac

caitlinadams committed:
Fix broken links

caitlinadams committed:
Update variable names for clarity and remove reference to .odc extension

caitlinadams committed:
Fix missing text in import section

caitlinadams committed:
Improve readability and consistency

caitlinadams committed:
update to api/indexed-data/datacube.index.abstract.rst

caitlinadams committed:
Add RunTimeWarnings to spellcheck

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

pjonsson commented on docs/tutorials/odc-stac.rst:
```diff
@@ -277,14 +282,19 @@ Type the following into the empty cell below the **Export loaded data** heading:
+   Make sure you download these files from the file browser before exiting the tutorial space as they will be deleted when the tutorial space is closed.
+
+   To download, open the file-browser by clicking the folder icon in the left menu bar.
+   Then, download the "LULC_year.tif" files by right-clicking each file and selecting "Download".
```
I think there should be something to indicate there isn't supposed to be a file called `LULC_year.tif`, perhaps call it `LULC_<year>.tif` instead?

caitlinadams replied:
Thanks, I've updated this.

pjonsson commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+
+   items = stac_client.search(
+       collections=desired_collections,
+       intersects=aoi_geometry,
```
Should the `desired_` prefix be added to the names of the other search parameters as well for consistency?

caitlinadams replied:
Yes, I've updated all search parameters.

pjonsson commented:
> The design pattern is that a full walkthrough is available in the docs page, and this is supported by notebooks held in a [companion repository](https://github.com/opendatacube/tutorial-odc-stac). This allows users to easily launch a binder instance and complete the tutorial. The companion repository is independently documented so that people can run it if accessing the tutorial via GitHub rather than the docs website.

@caitlinadams it feels like there could be other things that would be worthy of a tutorial as well, would it make sense to have one generic tutorial-repository with the odc-stac tutorial in some subdirectory, and other tutorials in other subdirectories?

(I keep doing the same change over and over in different repositories, and not having 25 tutorial repositories that have to be updated seem like a great time saver if any API changes in the future.)

caitlinadams commented:
> @caitlinadams it feels like there could be other things that would be worthy of a tutorial as well, would it make sense to have one generic tutorial-repository with the odc-stac tutorial in some subdirectory, and other tutorials in other subdirectories?
> 
> (I keep doing the same change over and over in different repositories, and not having 25 tutorial repositories that have to be updated seem like a great time saver if any API changes in the future.)

That's a very valid point! I'm inclined to agree, my only reservation is that Binder doesn't support this approach particularly well... and it's currently the simplest option for giving novice users a compute environment. The biggest barrier that I can see comes from Binder operating on a repository basis; it builds a Python environment from a single `requirements.txt` or `environment.yaml` file. My concern here is that we'd then need to manage an environment that works for all tutorials. I'd be open to trying this out -- we could always still have a specific environment file in each tutorial subfolder that users could build from if they want a local environment instead. 

I can have a chat with Brian about this and we'll make a plan for how we might test this.

In the long run, I would like to also be able to co-locate the docs `.rst` files with the tutorial source material, and then have that be built as part of the readthedocs page. Conversations I've had with Damien lead me to believe this is possible, but I haven't had the bandwidth to try and work it out.

pjonsson commented:
> > (I keep doing the same change over and over in different repositories, and not having 25 tutorial repositories that have to be updated seem like a great time saver if any API changes in the future.)
> 
> The biggest barrier that I can see comes from Binder operating on a repository basis; it builds a Python environment from a single `requirements.txt` or `environment.yaml` file. My concern here is that we'd then need to manage an environment that works for all tutorials.

I think there are options between the two extremes of one environment for everything and one environment for every single thing, the most obvious being one environment for most things and a couple of special environments for the exceptions.

I don't know anything about Binder, but I can say from experience that starting with one repository/environment and splitting that if/when the need arises in the future will be far less work than trying to merge a dozen tutorial repositories after realizing there should have only been one repository. There's also a possibility that Binder will support requirements in a folder by the time we have tutorials that require a different environment.

> In the long run, I would like to also be able to co-locate the docs `.rst` files with the tutorial source material, and then have that be built as part of the readthedocs page.

I don't know if anyone has looked at the documentation build, but my impression is that datacube-core has chosen a non-standard layout for the documentation directory for some reason, and that choice causes most of the wall of warnings seen in the readthedocs job. I don't see a lot of pull requests fixing documentation, so I'm quite tempted to make a PR that just moves the documentation into a standard layout so we can hopefully get rid of the wall of warnings.

Moving things around would obviously interfere with your forthcoming work of getting things to build together, should I try to prioritize that work so you get a starting point that is aligned with the assumptions in the readthedocs documentation?

caitlinadams commented:
> I don't know if anyone has looked at the documentation build, but my impression is that datacube-core has chosen a non-standard layout for the documentation directory for some reason, and that choice causes most of the wall of warnings seen in the readthedocs job. I don't see a lot of pull requests fixing documentation, so I'm quite tempted to make a PR that just moves the documentation into a standard layout so we can hopefully get rid of the wall of warnings.
> 
> Moving things around would obviously interfere with your forthcoming work of getting things to build together, should I try to prioritize that work so you get a starting point that is aligned with the assumptions in the readthedocs documentation?

I believe @omad did some tidy up with updating the template to improve the lay-out, but I'm not aware of any other background/work covering why our readthedocs structure is the way it is. I am in favour of moving to whatever readthedocs suggests as standard practice, and so would be very supportive of the PR you're suggesting. Perhaps worth raising in the technical subcommittee to see if anyone has concerns? 

I'm not able to prioritise the work on getting things to build together, so more than happy for you to work on a restructure if the rest of the technical subcommittee is supportive. I am also happy to assist with the restructure if there are clear tasks that can be delegated. Thanks @pjonsson!

caitlinadams committed:
Change tutorial AOI to resolve warnings from LULC product. Tidy up

caitlinadams committed:
Update wordlist.txt

pjonsson commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+============
+
+In this tutorial we will use Python libraries to find and load Land Use and Land Cover data from the freely available `Impact Observatory Annual Land Use Land Cover <https://planetarycomputer.microsoft.com/dataset/io-lulc-annual-v02>`_ product.
+After loading the data, we will export each year of data as a  Cloud Optimised GeoTIFF.
```
I think there is a double space before "Cloud".

pjonsson commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+The tutorial environment may take a few minutes to start.
+
+.. image:: https://mybinder.org/badge_logo.svg
+ :target: https://mybinder.org/v2/gh/opendatacube/tutorial-odc-stac/main?urlpath=%2Fdoc%2Ftree%2Ftutorial.ipynb
```
If we're going for one "common" tutorial repository to start with, I think it would make sense to generalize the repository name now before people start using it.

I shouldn't be the one making the decision, but my stance is that splitting the repository at a later stage will be easy (basically cut and paste) whereas contracting will be hard (discovering that people have cut and pasted differently so "identical" code has diverged, and aligning the code reveals a bunch of bugs that have to be addressed first). If we enter speculative territory, my gut feeling is that quite a few tutorials could manage with just odc-stac/odc-geo/odc-loader+supporting packages like pystac/pystac_client/etc., and all those packages should co-exist beautifully in a single environment.

If anyone has examples of packages that are problematic to put in the same environment, I'm happy to take a look at it. My optimistic view is that most software isn't problematic by design, it's just a time shortage among the contributors, and they usually welcome external contributions.

pjonsson commented on docs/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+ :alt: A view of a Jupyter notebook environment showing that the tutorial.ipynb file is loaded. The file is open with the header "odc-stac tutorial".
+ :align: center
+
+| We also recommend that you open the file browser by clicking the folder icon on the left-hand menu bar (circled in red in the image above). This will allow you to see the output files at the end of the tutorial.
```
I imagine this field isn't great for people who are red-green colorblind, but could a "near the top" (or similar) be inserted into the first sentence so it's blatantly obvious where one should look?

omad replied:
As someone who is red-green colourblind. I'd skimmed and missed the paragraph, and missed the circle in the image.
But.. it's not critical for understanding what's going on, so is fine.

pjonsson commented on wordlist.txt:
```diff
@@ -503,6 +524,7 @@ rtd
 rtype
 Ruff
 runtime
+RunTimeWarnings
```
I think this isn't used in the latest iteration of the tutorial.

pjonsson commented:
I don't know anything about the subject area, but I've read the latest iteration of the tutorial from start to finish and left some comments along the way, but on the whole, I think it looks good.

I also resolved a bunch of conversations that seem to have been addressed so we don't have to look at them.

I'll try to squeeze in moving the documentation around sometime early next week so you can rebase this branch on top of that so your name ends up in git blame for the tutorial.

caitlinadams committed:
Minor formatting and wording fixes for odc-stac tutorail

caitlinadams committed:
Merge branch 'develop' into doc-tutorial-odcstac

caitlinadams committed:
Update link to binder

caitlinadams committed:
Update wordlist

pjonsson reviewed (approved):
@caitlinadams I think the `_static` and `tutorials` directories need to go under `source` for the link from the "Tutorials" box on the index page to work, but other than that this looks great, thank you!

Edit: I support this but I don't think I should be the sole arbiter, so would be great if someone else also approves this.

caitlinadams committed:
Move materials under source

omad reviewed (changes requested):
This looks great!

However, there's a bunch of minor changes I want before we merge it.

omad commented on docs/_static/tutorial-images/odc-stac/aoi.png:
**Duplicates**
These images were copied instead of moved into the `_source/` directory structure.

We need to either rewrite the history or do a "Squash merge" when the PR is merged, if there's a new commit to remove them.

**Image location**
Although it's not necessarily followed in this docs. It's a better idea to keep images in the same directory (or very close by) to the document that's using them. It's not necessary to put them in `_static/`, and just ends up with a whole lot of bloat in `_static/` that gets hard to manage.

**More efficient image format**
I also think that we should use AVIF or WebP for images. PNG is really old and inefficient, 1MB for this image is too large. WebP and AVIF come in at 58KB and 64KB respectively.

WebP is [slightly older and more compatible](https://caniuse.com/webp) with a "Baseline - Widely available across major browsers". AVIF is newer and better but [slightly less widely supported](https://caniuse.com/avif) with a "Baseline 2024 - Newly available across major browsers" rating.

I'd be happy with either, leaning slightly towards AVIF.

caitlinadams replied:
I'm comfortable with doing a squash merge for this PR since it's a new document. I don't think there's history I'm really desperate to keep. I'm not sure how to edit the git history, so a squash merge seems the easiest approach.

I've updated the images and have a commit that deletes the old ones, but would appreciate you double checking I've caught everything.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+
+The notebook requires five libraries to run:
+
+* :code:`geopandas` for loading an area of interest from a GeoJSON file
```
Rather than plain text, I think linking to these libraries would be better.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+
+The area of interest is the island of La Gomera, one of the Canary Islands.
+
+.. image:: ../_static/tutorial-images/odc-stac/aoi.png
```
If we move the images this, we can change this image location. The `../_static/` is fragile in the case of changing the directory structure.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+Search for items
+^^^^^^^^^^^^^^^^
+
+After setting up the :code:`Client`, we use the :code:`search` method to find items that match our chosen collection, area of interest, and date range.
```
We can set up [intersphinx](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) links to go to the `Client` and `search()` API documentation for `pystac-client`.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+   print(desired_date_range)
+
+should return :code:`('2017-01-01', '2023-01-01')`.
+If you see a different date range, return to the **Set up query parameters - Date range** section and ensure your :code:`desired_start_date` and :code:`desired_end_date` values match those given in the instructions.
```
"Set up query parameters - Date range" should be a link to that part of the document.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+Load items with odc-stac
+------------------------
+
+After producing a list of items to load, we can use the :code:`load` function from :code:`odc-stac` to read the requested assets from the items and return them as xarrays.
```
More intersphinx link opportunities here.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
+
+After running the cell, you should see a :code:`xarray.Dataset` summary.
```
Intersphinx.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+---------------------
+
+To confirm that we have loaded the requested data, it is useful to visualise it. 
+We can use :code:`xarray`'s built-in plotting functionality to make a basic plot.
```
Intersphinx to the xarray plotting documentation page.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+After running the cell, you should see the following visualisation.
+The colours in the plot represent the following land cover classes:
+
+- dark blue: water
```
I found this list a little hard to read, it might work better as a table.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+
+Once you have loaded and checked your data, it is often useful to export it.
+This allows you to use the data in other software and analyses.
+The :code:`odc-geo` library provides the :code:`write_cog` function to generate and write these files from an :code:`xarray`.
```
Intersphinx.

omad commented on wordlist.txt:
```diff
@@ -199,13 +200,17 @@ ESRI
 executor
 f'file
 fd
+Fdoc
```
🤦🏻
I need to go and give ❤️ to https://github.com/opendatacube/datacube-core/pull/2097 so that we don't need to waste time on this file for this sort of mundane craziness.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+   import planetary_computer
+   from pystac_client import Client
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
```
The `code` role can be replaced with `kbd` as more semantically correct markup.

https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-kbd

```suggestion
When you have finished, run the cell by pressing :kbd:`Shift+Enter` on your keyboard.
```

omad commented:
To create AVIF images install `libavif` from https://github.com/AOMediaCodec/libavif. Then use like: `avifenc aoi.png aoi.avif`

To create WebP install `cwebp` following the instructions at https://developers.google.com/speed/webp/docs/using and then run: `cwebp aoi.png -o aoi.webp`

I found AVIF to give better colour saturation. WebP was a bit better with `cwebp -sharp_yuv aoi.png -o aoi_sharp_yuv.webp`.

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+   For this tutorial, we believe you will learn more if you type the code yourself, rather than using copy-paste.
+   Typing encourages you to slow down and think about what you're doing, which will help you gain understanding!
+
+   If you are stuck at any point, open the :code:`tutorial_solution.ipynb` notebook that is available in the file browser.
```
```suggestion
   If you are stuck at any point, open the :file:`tutorial_solution.ipynb` notebook that is available in the file browser.
```

https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+Area of interest
+^^^^^^^^^^^^^^^^
+
+We specify the area of interest using the :code:`aoi.geojson` file, which can be loaded with :code:`geopandas`.
```
```suggestion
We specify the area of interest using the :file:`aoi.geojson` file, which can be loaded with :code:`geopandas`.
```

https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-file

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+   desired_aoi = gpd.read_file("aoi.geojson")
+   desired_aoi_geometry = desired_aoi.iloc[0].geometry
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
```
```suggestion
When you have finished, run the cell by pressing :kbd:`Shift+Enter` on your keyboard.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+   desired_date_range = (desired_start_date, desired_end_date)
+
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
```
```suggestion
When you have finished, run the cell by pressing :kbd:`Shift+Enter` on your keyboard.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+   desired_collections = ["io-lulc-annual-v02"]
+   desired_assets = ["data"]
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
```
```suggestion
When you have finished, run the cell by pressing :kbd:`Shift+Enter` on your keyboard.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+      modifier=planetary_computer.sign_inplace,
+   )
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
```
```suggestion
When you have finished, run the cell by pressing :kbd:`Shift+Enter` on your keyboard.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+
+   print(f"Found {len(items)} items")
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
```
```suggestion
When you have finished, run the cell by pressing :kbd:`Shift+Enter` on your keyboard.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+
+   ds
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
```
```suggestion
When you have finished, run the cell by pressing :kbd:`Shift+Enter` on your keyboard.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+
+   ds["data"].plot.imshow(col="time", col_wrap=3)
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
```
```suggestion
When you have finished, run the cell by pressing :kbd:`Shift+Enter` on your keyboard.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+           overwrite=True,
+       )
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
```
```suggestion
When you have finished, run the cell by pressing :kbd:`Shift+Enter` on your keyboard.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+       )
+
+When you have finished, run the cell by pressing :code:`Shift+Enter` on your keyboard.
+You should see seven new files in the file browser, starting with "LULC_2017.tif" and ending with "LULC_2023.tif".
```
```suggestion
You should see seven new files in the file browser, starting with :file:`LULC_2017.tif` and ending with :file:`LULC_2023.tif`.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+   Make sure you download these files from the file browser before exiting the tutorial space as they will be deleted when the tutorial space is closed.
+
+   To download, open the file-browser by clicking the folder icon in the left menu bar.
+   Then, download the Land Use Land Cover file for each year (denoted as `LULC_<year>.tif`) by right-clicking each file and selecting "Download".
```
```suggestion
   Then, download the Land Use Land Cover file for each year (denoted as :file:`LULC_{year}.tif`) by right-clicking each file and selecting :guilabel:`Download`.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+
+We specify the area of interest using the :code:`aoi.geojson` file, which can be loaded with :code:`geopandas`.
+
+The area of interest is the island of La Gomera, one of the Canary Islands.
```
```suggestion
The area of interest is the island of `La Gomera <https://en.wikipedia.org/wiki/La_Gomera>`_, one of the `Canary Islands <https://en.wikipedia.org/wiki/Canary_Islands>`_.
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+  * what (data provider and product)
+  * where (area of interest)
+  * when (date range)
+* Use `pystac-client`_ to connect to a Spatio-Temporal Asset Catalog (STAC) 
```
```suggestion
* Use `pystac-client`_ to connect to a `Spatio-Temporal Asset Catalog (STAC) <https://stacspec.org/en>`_ 
```

omad commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,300 @@
+============
+
+In this tutorial we will use Python libraries to find and load Land Use and Land Cover data from the freely available `Impact Observatory Annual Land Use Land Cover <https://planetarycomputer.microsoft.com/dataset/io-lulc-annual-v02>`_ product.
+After loading the data, we will export each year of data as a Cloud Optimised GeoTIFF.
```
```suggestion
After loading the data, we will export each year of data as a `Cloud Optimised GeoTIFF <https://cogeo.org/>`_.
```

caitlinadams committed:
Update RST roles on multiple commands as suggested

caitlinadams committed:
Add intersphinx links to config and try using in odc-stac tutorial

caitlinadams committed:
Remove unused links and update wordlist

caitlinadams committed:
Change intersphinx references to doc role

caitlinadams committed:
Add intersphinx links and move images

caitlinadams committed:
Move images to .avif format

caitlinadams committed:
remove unused older tutoral png files

caitlinadams committed:
Add avif to wordlist

pjonsson commented on docs/source/tutorials/odc-stac.rst:
```diff
@@ -0,0 +1,306 @@
+      bands=desired_assets,
+      geopolygon=aoi_geometry,
+      crs="utm",
+      resolution=30
```
Just noticed that all the other multi-line calls have trailing comma after the last parameter, but this one is missing a trailing comma.

caitlinadams replied:
Thanks for picking this up. I've added the comma

caitlinadams committed:
Add comma to odc-stac load instruction

omad reviewed (changes requested):
This is great, thank you @caitlinadams !

I'll merge it in a few minutes, I'm going to simplify a few of the Python API links first though.

At this stage, I plan to do a squash merge to only include the current versions of files in the repo history, to not add bloat with the different image formats that have been removed.

Is there any history you would like to have kept? Speak up if there is and I can do some rebasing/history pruning if you'd prefer.

caitlinadams commented:
> I'll merge it in a few minutes, I'm going to simplify a few of the Python API links first though.

Awesome, thank you!

> At this stage, I plan to do a squash merge to only include the current versions of files in the repo history, to not add bloat with the different image formats that have been removed.
> Is there any history you would like to have kept? Speak up if there is and I can do some rebasing/history pruning if you'd prefer.

Very happy to go ahead with the squash merge, there's nothing that I need to keep, and I went through a few iterations of various things, so quite happy to just have one clean commit.

omad committed:
Simplify intersphinx API links

It's not necessary to specifying the source documentation site, Sphinx
is smart enough to find references to fully qualified Python objects.

There's also a neat syntax for shortening the link text to the last
component of the target, by prefixing the target with a tilde '~'.

See https://www.sphinx-doc.org/en/master/usage/referencing.html#syntax
for details.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

omad commented:
Ugh, that took waaay longer than I intended, due to a deep rabbit whole with the `autosummary` directive putting generated files into the `source/` tree.

I've committed the two little simplifications I wanted:
- It's not necessary to specifying the source documentation site, Sphinx
is smart enough to find references to fully qualified Python objects.
- There's also a neat syntax for shortening the link text to the last
component of the target, by prefixing the target with a tilde `~`.

  See https://www.sphinx-doc.org/en/master/usage/referencing.html#syntax
for details.

The other thing I'd note is about the section on writing the data back out to COG. It's very not clearly documented in the `odc-geo` docs, but, while the methods in `odc.geo.xr` can be imported an used directly, that's not the only way! They're also made available as methods on xarray DataSets and DataArrays.

eg.
```python
# this line makes the odc xarray extensions available
import odc.geo.xr

# After which, say you have a DataArray like in the tutorial, you can run
ds_single_year.odc.write_cog(
    f"LULC_{years[timestep]}.tif",
    overwrite=True,
)
```

Written like that, it's pretty insignificant a change. But when you're writing code in a notebook, it can be quite handy to do `myds.odc.<tab>` and see the options, and being able to  display a `ds.odc.geobox` or quickly make a folium map with `ds.odc.explore()`.

It's sort of covered on https://odc-geo.readthedocs.io/en/stable/intro-xr.html and https://odc-geo.readthedocs.io/en/stable/raster-tools.html , but, also sort of not and a lot of people probably aren't aware.

caitlinadams commented:
Thank you @SpacemanPaul, @robbibt, @pjonsson, @Kirill888 and @omad for your support on this PR! I'm delighted to have our first official tutorial done!
